### ISSUE TITLE
Undo and Redo problems on OSX in Typescript files with Auto Save enabled

### ISSUE THREAD

**Issue by sek** [Microsoft/vscode#49567]
Issue Type: <b>Bug</b>

### Steps to Reproduce

1. Enable Auto Save in the File menu
2. Create a typescript file - ex "test-file.ts" and open in editor
3. Put some words in the file - ex: "// here is a comment with some words in it"
4. Delete one word - ex: comment
5. Delete a second word - ex: some
6. Undo each deletion with Command-Z or Edit->Undo
7. Redo deleting each word with Command-Shift-Z or Edit->Redo

### Expected

- Undo twice should restore both deleted words
- Redo twice should re-delete both deleted words

### Actual

- Multiple undo actions are required to undo each deletion - in three tests, I had to press Command-Z between 4 and 6 times to get the 2 deleted words to reappear
- Redo does not work at all after undoing the 2 deletions

### Additional Details

- If "Auto Save" is disabled, the scenario works as expected (2 Undo Actions, 2 Redo Actions)
- If editing a text file, even with "Auto Save" enabled, it works as expected (test-file.text with identical contents and undo/redos behaves exactly as I'd hope test-file.ts would behave)
- This is partially reproducible with manual saves in a Typescript file, ex: if manually saving after step 6, redo is no longer possible

Also see: https://github.com/Microsoft/vscode/issues/26228 for an earlier report of the problem - closed as "caused by extension", but perhaps disabling "Auto Save" was the real "fix"

VS Code version: Code 1.22.2 (3aeede733d9a3098f7b4bdc1f66b63b0f48c1ef9, 2018-04-12T16:32:53.389Z)
OS version: Darwin x64 17.4.0

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-4578U CPU @ 3.00GHz (4 x 3000)|
|Load (avg)|3, 3, 5|
|Memory (System)|16.00GB (1.75GB free)|
|Process Argv|/Applications/Visual Studio Code.app/Contents/MacOS/Electron|
|Screen Reader|no|
|VM|0%|

</details>Extensions: none

**Comment by alexandrudima** [comment]
@sek Are you using trim trailing whitespace on save ?

Anytime you would save, an edit would be pushed that trims trailing whitespace, which would explain what you are describing. Would you expect trim trailing whitespace to be disabled when there is something in the redo stack ?

**Comment by sek** [comment]
Hi @alexandrudima,

I am using format on save, so I thought perhaps that might be creating additional edits as you suspect.  However, I wasn't able to change the behavior by changing the formatOnSave options (I commented all workspace and user settings)

Then I played around some more, disabled all extensions again, and found that I couldn't reproduce.  Strange that VS Code even reported all extensions disabled the other day, the Sort Typescript import statements must have been still enabled, because that's the one I narrowed it down to:
https://marketplace.visualstudio.com/items?itemName=miclo.sort-typescript-imports
https://github.com/MLoughry/sort-typescript-imports/

With it disabled, I don't get the behavior where I have to undo multiple times for each edit (I just experienced 5 Command-Z presses to undo a single edit)

Redo stack is only cleared if a formatOnSave change is made, but not if no change is made - so that seems about as good as it gets.

So it seems the issue should be closed here, and I'll see about following up on that repository

The two things I think should be fixed with the Sort Typescript Imports plugin:

- It should only run on manual save - not auto-save  (like editor.formatOnSave)
- It should only create edits if the imports are actually changed, not if the file is identical before and after sorting

Any quick tips on those points?

Thanks!
